preprocessing-service/preprocess.py

```diff
@@ -1,32 +1,72 @@
 # Standard Library
 import asyncio
+import json
 import logging
+import os
+import sys
 import time
 
 # Third Party
+from elasticsearch import AsyncElasticsearch
 from opni_proto.log_anomaly_payload_pb import Payload, PayloadList
 import pandas as pd
 from masker import LogMasker
 from opni_nats import NatsWrapper
 
 logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(name)s - %(message)s")
-
+workload_parameters_dict = dict()
 nw = NatsWrapper()
 
+ES_ENDPOINT = os.environ["ES_ENDPOINT"]
+ES_USERNAME = os.getenv("ES_USERNAME", "admin")
+ES_PASSWORD = os.getenv("ES_PASSWORD", "admin")
+
+es_instance = AsyncElasticsearch(
+    [ES_ENDPOINT],
+    port=9200,
+    http_auth=(ES_USERNAME, ES_PASSWORD),
+    verify_certs=False,
+    use_ssl=True,
+)
+
 async def consume_logs(mask_logs_queue):
-    async def subscribe_handler(msg):
+    async def pretrained_subscribe_handler(msg):
         payload_data = msg.data
         log_payload = Payload()
         await mask_logs_queue.put(log_payload.parse(payload_data))
 
+    async def workload_parameters_handler(msg):
+        global workload_parameters_dict
+        workload_parameters_dict = json.loads(msg.data.decode())
 
     await nw.subscribe(
         nats_subject="raw_logs",
         nats_queue="workers",
         payload_queue=mask_logs_queue,
-        subscribe_handler=subscribe_handler,
+        subscribe_handler=pretrained_subscribe_handler,
+    )
+
+    await nw.subscribe(
+        nats_subject="workload_parameters",
+        nats_queue="workers",
+        subscribe_handler=workload_parameters_handler,
     )
 
+async def get_latest_workload():
+    global workload_parameters_dict
+    query_body = {"sort": [{"time": {"order": "desc"}}],"query": {"match_all": {}}}
+    try:
+        latest_workload = await es_instance.search(index="model-training-parameters",body=query_body,size=1)
+        workload_parameters_dict = json.loads(latest_workload["hits"]["hits"][0]["_source"]["parameters"])
```

Comment:
does latest_workload["hits"]["hits"][0] give you the latest record or oldest?

Reply:
I'm sorting the time in descending order so the first record will be the latest record.